exec: add nulls support to vectorized merge joiner

georgeutsin opened the pull request:
**exec: add null support to the vectorized merge joiner**
Added support for nulls in both equality and non equality columns
for the vectorized merge joiner.

To do this, the joiner templating had to be refactored to handle
the different combinations of nulls in both the probing and
building phases.

There's a slight performance hit due to the added complexity, so
the benchmarks are shown below.

```
name                                      old time/op    new time/op    delta
MergeJoiner/rows=1024-8                     38.9µs ± 6%    39.8µs ± 1%     ~     (p=0.123 n=10+10)
MergeJoiner/rows=4096-8                      148µs ± 4%     152µs ± 0%   +3.21%  (p=0.002 n=10+9)
MergeJoiner/rows=16384-8                     575µs ± 3%     588µs ± 1%   +2.32%  (p=0.043 n=10+10)
MergeJoiner/rows=1048576-8                  35.1ms ± 2%    36.2ms ± 1%   +3.20%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8       38.1µs ± 2%    40.0µs ± 0%   +5.04%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=4096-8        145µs ± 0%     153µs ± 0%   +5.72%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8       561µs ± 1%     589µs ± 0%   +4.89%  (p=0.000 n=10+8)
MergeJoiner/oneSideRepeat-rows=1048576-8    34.8ms ± 6%    35.9ms ± 1%   +3.23%  (p=0.002 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8     39.0µs ± 1%    42.5µs ± 1%   +9.03%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      185µs ± 1%     191µs ± 0%   +3.13%  (p=0.000 n=10+9)
MergeJoiner/bothSidesRepeat-rows=16384-8    1.44ms ± 1%    1.46ms ± 0%   +1.13%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=32768-8    5.10ms ± 1%    5.12ms ± 1%     ~     (p=0.529 n=10+10)

name                                      old speed      new speed      delta
MergeJoiner/rows=1024-8                   1.69GB/s ± 5%  1.65GB/s ± 1%     ~     (p=0.123 n=10+10)
MergeJoiner/rows=4096-8                   1.78GB/s ± 4%  1.72GB/s ± 0%   -3.14%  (p=0.002 n=10+9)
MergeJoiner/rows=16384-8                  1.82GB/s ± 3%  1.78GB/s ± 1%   -2.30%  (p=0.043 n=10+10)
MergeJoiner/rows=1048576-8                1.91GB/s ± 2%  1.85GB/s ± 1%   -3.10%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8     1.72GB/s ± 2%  1.64GB/s ± 0%   -4.81%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=4096-8     1.81GB/s ± 0%  1.71GB/s ± 0%   -5.41%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8    1.87GB/s ± 1%  1.78GB/s ± 0%   -4.66%  (p=0.000 n=10+8)
MergeJoiner/oneSideRepeat-rows=1048576-8  1.93GB/s ± 6%  1.87GB/s ± 1%   -3.18%  (p=0.002 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8   1.68GB/s ± 1%  1.54GB/s ± 1%   -8.28%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8   1.42GB/s ± 1%  1.37GB/s ± 1%   -3.12%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=16384-8   726MB/s ± 1%   718MB/s ± 0%   -1.12%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=32768-8   411MB/s ± 1%   410MB/s ± 1%     ~     (p=0.529 n=10+10)

name                                      old alloc/op   new alloc/op   delta
MergeJoiner/rows=1024-8                      6.20B ±45%     5.00B ± 0%     ~     (p=0.137 n=10+8)
MergeJoiner/rows=4096-8                      27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/rows=16384-8                     90.0B ± 0%     90.0B ± 0%     ~     (all equal)
MergeJoiner/rows=1048576-8                  5.45kB ± 0%    5.45kB ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1024-8        5.00B ± 0%     7.80B ±36%  +56.00%  (p=0.005 n=9+10)
MergeJoiner/oneSideRepeat-rows=4096-8        27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=16384-8       90.0B ± 0%     90.0B ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1048576-8    5.45kB ± 0%    5.45kB ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=1024-8      5.00B ± 0%     9.00B ± 0%  +80.00%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=16384-8      272B ± 0%      272B ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=32768-8      907B ± 0%      907B ± 0%     ~     (all equal)

name                                      old allocs/op  new allocs/op  delta
MergeJoiner/rows=1024-8                       0.00           0.00          ~     (all equal)
MergeJoiner/rows=4096-8                       0.00           0.00          ~     (all equal)
MergeJoiner/rows=16384-8                      0.00           0.00          ~     (all equal)
MergeJoiner/rows=1048576-8                    1.00 ± 0%      1.00 ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1024-8         0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=4096-8         0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=16384-8        0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1048576-8      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=1024-8       0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=4096-8       0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=16384-8      0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=32768-8      0.00           0.00          ~     (all equal)
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/36727)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Thanks for the quick work on this!

Also, there are two commits here, so don't forget to squash 'em.

> Reviewed 1 of 5 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-:-Lc93c_d7EQmwqMn5lJf:b-ry2ie9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @georgeutsin, and @jordanlewis)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8wPsk3baaTmTAoYIO:-Lc8wPsk3baaTmTAoYIP:bldc0um) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/distsqlrun/column_exec_setup.go#L287)):*
> ```Go
> 		nRightCols := uint32(len(rightTypes))
> 
> 		leftEqCols := make([]uint32, 0, nLeftCols)
> ```

I think now that we're using both `ColIdx` and `Direction` it might make sense to use `distsqlpb.Ordering_Column` directly.
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 290 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8ucXM99QwWyQUVNFO:-Lc8ucXM99QwWyQUVNFP:b-wtrn2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/distsqlrun/column_exec_setup.go#L290)):*
> ```Go
> 		leftDirections := make([]distsqlpb.Ordering_Column_Direction, 0, nLeftCols)
> 		rightEqCols := make([]uint32, 0, nRightCols)
> 		rightDirections := make([]distsqlpb.Ordering_Column_Direction, 0, nLeftCols)
> ```

`nRightCols`.
___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 170 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8vEOs1N1StDNO615T:-Lc8vEOs1N1StDNO615U:b6jsaqj) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/distsqlrun/columnar_operators_test.go#L170)):*
> ```Go
> }
> 
> // generateColumnOrdering produces a random ordering of nOrderingCols columns
> ```

[nit]: function name has changed, so the comment needs adjusting (as well as the panic message below). Also, I'd mention that directions are fixed in the beginning of the comment like `generateColumnOrderingWithDirections, given a slice of column directions, produces a ...`.
___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 199 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8vblg4K3Tid6RaXKC:-Lc8vblg4K3Tid6RaXKD:b-dgf7vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/distsqlrun/columnar_operators_test.go#L199)):*
> ```Go
> }
> 
> // generateColumnOrdering is a wrapper for generateColumnOrderingWithDirections
> ```

[nit]: I'd try not to mention the "details" of the method here but the "purpose" instead (maybe the comment as it was on the method with this name with an explicit mention of randomly generated directions).
___
*[pkg/sql/exec/mergejoiner.go, line 127 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8x7ks3u0OxrBQBB2h:-Lc8x7ks3u0OxrBQBB2i:brt4smb) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner.go#L127)):*
> ```Go
> 
> 	// directions specifies the ordering direction of each column. Note that each
> 	// direction corresponds to an equality column at the same location, ie the
> ```

[super nit]: `i.e.` Also, this comment has another argument (that both slices are kept "in sync") for using `Ordering_Column` directly.
___
*[pkg/sql/exec/mergejoiner_test.go, line 802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8y2JP-Ts8TDfUG-vm:-Lc8y2JP-Ts8TDfUG-vn:binxgbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_test.go#L802)):*
> ```Go
> 			rightDirections: []distsqlpb.Ordering_Column_Direction{distsqlpb.Ordering_Column_ASC},
> 		},
> 	}
> ```

You should add a few sanity checks with DESC direction as well.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc9-BeQCyQl7a7Ir5nc:-Lc9-BeQCyQl7a7Ir5nd:b-1yw55c) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L99)):*
> ```Go
> 
> // {{/*
> func _PROBE_SWITCH(sel perm, asc bool, lNull bool, rNull bool) { // */}}
> ```

[question]: what is this `perm` type?
[nit]: I'd call these `lHasNulls` and `rHasNulls` for clarity.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8zRXE5CsgYsbwT0xW:-Lc8zRXE5CsgYsbwT0xX:b-u0lpt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L103)):*
> ```Go
> 	// {{ $sel := $.Sel }}
> 	switch colType {
> 	// {{range $.Global.MJOverloads }}
> ```

[question]: what is the point of this change?
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8yk0vBJJHLz_x1JGk:-Lc8yk0vBJJHLz_x1JGl:bv5jip3) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L115)):*
> ```Go
> 			// Expand or filter each group based on the current equality column.
> 			for curLIdx < curLLength && curRIdx < curRLength {
> 				// TODO (georgeutsin) change null check logic for non INNER joins
> ```

[nit]: all TODO comments I've seen are of the form `TODO(yuzefovich): ...` (also, missing period).
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc903y69509jm0RVVQc:-Lc903y69509jm0RVVQd:b90uwe4) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L231)):*
> ```Go
> 			if lVec.HasNulls() {
> 				if rVec.HasNulls() {
> 					_PROBE_SWITCH(_SEL_ARG, true, true, true)
> ```

[nit]: it's probably worth spelling out parameter names for the booleans.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc91CLu6W7xMaeckUxs:-Lc91CLu6W7xMaeckUxt:b5sz5be) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L266)):*
> ```Go
> 
> // {{/*
> func _LEFT_SWITCH(hasNulls bool, isSel bool) { // */}}
> ```

[nit]: maybe for consistency with `_PROBE_SWITCH` switch the order of arguments?
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90EhLD7_FiJvn4V2D:-Lc90EhLD7_FiJvn4V2E:b-r9zez7) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L272)):*
> ```Go
> 	// {{range $.Global.MJOverloads }}
> 	case _TYPES_T:
> 		// {{ if $.IsSel  }}
> ```

I don't understand this change. It seems like `if` case is empty, and `else` case will define local variables that are not used.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90Z182hu_8GMrudKz:-Lc90Z182hu_8GMrudL-:b-8rds1n) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L293)):*
> ```Go
> 
> 						// {{ if $.HasNulls }}
> 						// TODO (georgeutsin): create a SetNullRange(start, end) function in coldata.Nulls, and place this outside the tight loop.
> ```

[nit]: strive for 80 character limit per line comments.
___
*[pkg/sql/exec/execgen/cmd/execgen/mergejoiner_gen.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc92agyCco14CKijO8B:-Lc92agyCco14CKijO8C:b-c097bm) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/execgen/cmd/execgen/mergejoiner_gen.go#L109)):*
> ```Go
> 	// Create each permutation of selection vector state.
> 	selPermutations := make([]selPermutation, 4)
> 	i := 0
> ```

I prefer the old way - it was verbose, but there was nothing complicated about it. Not a strong opinion though.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc93I6r2sBIxysm4ZPR:-Lc93I6r2sBIxysm4ZPS:b-6ya61t) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L9)):*
> ```
> 
> statement ok
> INSERT INTO t1 VALUES (-1, -1), (0, 4), (2, 1), (3, 4), (5, 4);
> ```

[nit]: semicolon is unnecessary.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc93N63AxwTbgk2VFWs:-Lc93N63AxwTbgk2VFWt:b-8u05ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L45)):*
> ```
> 
> statement ok
> CREATE INDEX on t1 (k);
> ```

[question]: do we need to explicitly create an INDEX here? Isn't having a column defined as `PRIMARY KEY` on the schema enough?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
I think I got a bit confused with the commits. It seems like the second one is only testing related? Might make sense to just squash them as yahor says.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-:-LcBcjds0AR-dwIkbcPm:b-al9lm1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 19 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcBaRk3-Q8Fsi3vTz8Q:-LcBaRk3-Q8Fsi3vTz8R:b-7w0tq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/distsqlrun/columnar_operators_test.go#L19)):*
> ```Go
> import (
> 	"context"
> 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
> ```

I think these changes should be in the second commit, right?
___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcB_2cz3DMrpR4fMBWV:-LcB_2cz3DMrpR4fMBWW:brphcab) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/distsqlrun/columnar_operators_test.go#L69)):*
> ```Go
> 	var da sqlbase.DatumAlloc
> 	st := cluster.MakeTestingClusterSettings()
> 	evalCtx := tree.MakeTestingEvalContext(st)
> ```

What I meant on the previous PR was that if you don't use `st` below you might as well just `tree.MakeTestingEvalContext(cluster.MakeTestingClusterSettings)` but this is just a nit
___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcBbUnl0y5QrpCJ1lO9:-LcBbUnl0y5QrpCJ1lOA:b-i3gmq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/distsqlrun/columnar_operators_test.go#L137)):*
> ```Go
> 		rOrderingCols := generateColumnOrderingWithDirections(rng, nCols, nCols, directions)
> 
> 		lRows := sqlbase.MakeRandIntRowsInRange(rng, nRows, nCols, maxNum)
> ```

Can this generate nulls?
___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcB_cS18zYWQBrW9OGB:-LcB_cS18zYWQBrW9OGC:bkea4hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/distsqlrun/columnar_operators_test.go#L173)):*
> ```Go
> // on a table with nCols columns, so nOrderingCols must be not greater than
> // nCols, given a slice of column directions.
> func generateColumnOrderingWithDirections(
> ```

I don't think I see the need for `generateColumnOrdering` to be split up. It seems the only time you call `generateColumnOrderingWithDirections` is with `generateColumnDirections` which is what `generateColumnOrdering` does. I would prefer to just have it as before (minus the `randomizeDirection`) boolean and simply overwrite the ordering you get for the right columns with whatever you got for the left columns.
___
*[pkg/sql/exec/mergejoiner_test.go, line 802 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8y2JP-Ts8TDfUG-vm:-LcBcHTm1QyAA6I_hbyi:b97g3rh) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_test.go#L802)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

You should add a few sanity checks with DESC direction as well.
</blockquote></details>

Another thing to consider is to set these directions by default. You would do that by defining a `testCase` struct with an init method that is called before `runTests` that would populate any missing directions with `distsqlpb.Ordering_Column_ASC`
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 39 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcBbCAY9BRNPgZO4NJ5:-LcBbCAY9BRNPgZO4NJ6:b8j5xo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L39)):*
> ```
> 
> statement ok
> CREATE TABLE  t1 (k INT)
> ```

nit: extra space between `TABLE` and `t1`


<!-- Sent from Reviewable.io -->

georgeutsin reviewed (commented):
In all honesty, these are two unrelated changes that happened to both be touching the same parts of the code. I think I'll just open separate PRs for them

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-:-LcBjCz6AWKxalUiZlFS:b-cw3d8z)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)



<!-- Sent from Reviewable.io -->

georgeutsin committed:
exec: add null support to the vectorized merge joiner

Added support for nulls in both equality and non equality columns
for the vectorized merge joiner.

To do this, the joiner templating had to be refactored to handle
the different combinations of nulls in both the probing and
building phases.

There's a slight performance hit due to the added complexity, so
the benchmarks are shown below.

```
name                                      old time/op    new time/op    delta
MergeJoiner/rows=1024-8                     38.9µs ± 6%    39.8µs ± 1%     ~     (p=0.123 n=10+10)
MergeJoiner/rows=4096-8                      148µs ± 4%     152µs ± 0%   +3.21%  (p=0.002 n=10+9)
MergeJoiner/rows=16384-8                     575µs ± 3%     588µs ± 1%   +2.32%  (p=0.043 n=10+10)
MergeJoiner/rows=1048576-8                  35.1ms ± 2%    36.2ms ± 1%   +3.20%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8       38.1µs ± 2%    40.0µs ± 0%   +5.04%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=4096-8        145µs ± 0%     153µs ± 0%   +5.72%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8       561µs ± 1%     589µs ± 0%   +4.89%  (p=0.000 n=10+8)
MergeJoiner/oneSideRepeat-rows=1048576-8    34.8ms ± 6%    35.9ms ± 1%   +3.23%  (p=0.002 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8     39.0µs ± 1%    42.5µs ± 1%   +9.03%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      185µs ± 1%     191µs ± 0%   +3.13%  (p=0.000 n=10+9)
MergeJoiner/bothSidesRepeat-rows=16384-8    1.44ms ± 1%    1.46ms ± 0%   +1.13%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=32768-8    5.10ms ± 1%    5.12ms ± 1%     ~     (p=0.529 n=10+10)

name                                      old speed      new speed      delta
MergeJoiner/rows=1024-8                   1.69GB/s ± 5%  1.65GB/s ± 1%     ~     (p=0.123 n=10+10)
MergeJoiner/rows=4096-8                   1.78GB/s ± 4%  1.72GB/s ± 0%   -3.14%  (p=0.002 n=10+9)
MergeJoiner/rows=16384-8                  1.82GB/s ± 3%  1.78GB/s ± 1%   -2.30%  (p=0.043 n=10+10)
MergeJoiner/rows=1048576-8                1.91GB/s ± 2%  1.85GB/s ± 1%   -3.10%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=1024-8     1.72GB/s ± 2%  1.64GB/s ± 0%   -4.81%  (p=0.000 n=10+9)
MergeJoiner/oneSideRepeat-rows=4096-8     1.81GB/s ± 0%  1.71GB/s ± 0%   -5.41%  (p=0.000 n=10+10)
MergeJoiner/oneSideRepeat-rows=16384-8    1.87GB/s ± 1%  1.78GB/s ± 0%   -4.66%  (p=0.000 n=10+8)
MergeJoiner/oneSideRepeat-rows=1048576-8  1.93GB/s ± 6%  1.87GB/s ± 1%   -3.18%  (p=0.002 n=10+10)
MergeJoiner/bothSidesRepeat-rows=1024-8   1.68GB/s ± 1%  1.54GB/s ± 1%   -8.28%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8   1.42GB/s ± 1%  1.37GB/s ± 1%   -3.12%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=16384-8   726MB/s ± 1%   718MB/s ± 0%   -1.12%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=32768-8   411MB/s ± 1%   410MB/s ± 1%     ~     (p=0.529 n=10+10)

name                                      old alloc/op   new alloc/op   delta
MergeJoiner/rows=1024-8                      6.20B ±45%     5.00B ± 0%     ~     (p=0.137 n=10+8)
MergeJoiner/rows=4096-8                      27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/rows=16384-8                     90.0B ± 0%     90.0B ± 0%     ~     (all equal)
MergeJoiner/rows=1048576-8                  5.45kB ± 0%    5.45kB ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1024-8        5.00B ± 0%     7.80B ±36%  +56.00%  (p=0.005 n=9+10)
MergeJoiner/oneSideRepeat-rows=4096-8        27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=16384-8       90.0B ± 0%     90.0B ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1048576-8    5.45kB ± 0%    5.45kB ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=1024-8      5.00B ± 0%     9.00B ± 0%  +80.00%  (p=0.000 n=10+10)
MergeJoiner/bothSidesRepeat-rows=4096-8      27.0B ± 0%     27.0B ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=16384-8      272B ± 0%      272B ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=32768-8      907B ± 0%      907B ± 0%     ~     (all equal)

name                                      old allocs/op  new allocs/op  delta
MergeJoiner/rows=1024-8                       0.00           0.00          ~     (all equal)
MergeJoiner/rows=4096-8                       0.00           0.00          ~     (all equal)
MergeJoiner/rows=16384-8                      0.00           0.00          ~     (all equal)
MergeJoiner/rows=1048576-8                    1.00 ± 0%      1.00 ± 0%     ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1024-8         0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=4096-8         0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=16384-8        0.00           0.00          ~     (all equal)
MergeJoiner/oneSideRepeat-rows=1048576-8      1.00 ± 0%      1.00 ± 0%     ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=1024-8       0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=4096-8       0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=16384-8      0.00           0.00          ~     (all equal)
MergeJoiner/bothSidesRepeat-rows=32768-8      0.00           0.00          ~     (all equal)
```

Release note: None

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc9-BeQCyQl7a7Ir5nc:-LcByBoJ5HUEr65eCeK5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L99)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[question]: what is this `perm` type?
[nit]: I'd call these `lHasNulls` and `rHasNulls` for clarity.
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8zRXE5CsgYsbwT0xW:-LcByCsd3LkEbvIolUkg:blyvc3q) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L103)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[question]: what is the point of this change?
</blockquote></details>

Since we need a reference to the original `.` from the parent template, we copy it into Global so that we still have access to it, and thus access to the overloads from the "global" context. When we step into this function, the . changes to be all of the arguments that we pass in to the template (in the call to the template in probeBodyLSel_IS_L_SELRSel_IS_R_SEL)
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8yk0vBJJHLz_x1JGk:-LcByn7345zUdsc2Esyk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L115)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: all TODO comments I've seen are of the form `TODO(yuzefovich): ...` (also, missing period).
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc903y69509jm0RVVQc:-LcByvue-Sl816vwWzwg:bv60429) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L231)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it's probably worth spelling out parameter names for the booleans.
</blockquote></details>

Not sure how comments would interact with the regex for the template, since this line is going to be regexed and replaced with `{{ template probeSwitch $sel ...}}`
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 266 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc91CLu6W7xMaeckUxs:-LcBzwhZ3iP1KRlhFK7c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L266)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: maybe for consistency with `_PROBE_SWITCH` switch the order of arguments?
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90EhLD7_FiJvn4V2D:-LcBzzAQ03eLhKwwkRMM:b-dc96ck) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L272)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I don't understand this change. It seems like `if` case is empty, and `else` case will define local variables that are not used.
</blockquote></details>

I actually had to include the else cause specifically because the local variables aren't used in the `if` case, and I couldn't find a good way to negate variables in the template language
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 293 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90Z182hu_8GMrudKz:-LcC-RTd3fMx5ce9FPgc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L293)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: strive for 80 character limit per line comments.
</blockquote></details>

Done.
___
*[pkg/sql/exec/execgen/cmd/execgen/mergejoiner_gen.go, line 109 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc92agyCco14CKijO8B:-LcC-TfJ2sH2wijrRoXj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/execgen/cmd/execgen/mergejoiner_gen.go#L109)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I prefer the old way - it was verbose, but there was nothing complicated about it. Not a strong opinion though.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc93I6r2sBIxysm4ZPR:-LcC-WxFD72Jy4oDYPop:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L9)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: semicolon is unnecessary.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 45 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc93N63AxwTbgk2VFWs:-LcC-hzP86jyEH7Bj69t:b-puvbl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L45)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[question]: do we need to explicitly create an INDEX here? Isn't having a column defined as `PRIMARY KEY` on the schema enough?
</blockquote></details>

You're right, I'll remove it


<!-- Sent from Reviewable.io -->

georgeutsin commented:
RFAL :eyes: @asubiotto @yuzefovich, I split the PR up so now this is just nulls support.

Maybe @solongordon too

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

I was ok with having two commits in one PR, just squashing them before merging is - I think - our policy. Now when you open a different PR for the directions, please address the comments that are left on this PR but related to directions PR in there as well. In general, it is a good idea though to have two separate PRs.

> Reviewed 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-:-LcCFp5o7ryGFqEJpNdH:b-rqmjq7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 103 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8zRXE5CsgYsbwT0xW:-LcCGlhh6vdFIWIoEgTj:b-54avio) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L103)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

Since we need a reference to the original `.` from the parent template, we copy it into Global so that we still have access to it, and thus access to the overloads from the "global" context. When we step into this function, the . changes to be all of the arguments that we pass in to the template (in the call to the template in probeBodyLSel_IS_L_SELRSel_IS_R_SEL)
</blockquote></details>

Ok, thanks.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 115 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc8yk0vBJJHLz_x1JGk:-LcCHOE990VZTyug7iIs:bsj6see) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L115)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

[nit]: no space between TODO and parentheses.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 231 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc903y69509jm0RVVQc:-LcCHdHr7wJqUx72qEiG:b-f06wdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L231)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

Not sure how comments would interact with the regex for the template, since this line is going to be regexed and replaced with `{{ template probeSwitch $sel ...}}`
</blockquote></details>

Ok, makes sense, please disregard my comment.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90EhLD7_FiJvn4V2D:-LcCIMiY8Mxsu_EnIxqS:b-2u69l) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L272)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

I actually had to include the else cause specifically because the local variables aren't used in the `if` case, and I couldn't find a good way to negate variables in the template language
</blockquote></details>

Oh, I think I get it. Please leave a comment to describe the purpose of this if/else.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 9 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc93I6r2sBIxysm4ZPR:-LcCK8fk9LRuqEWEbhwl:b36h2n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L9)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

There are a few others in this file.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcCKCXsA0qoki61vI5i:-LcCKCXsA0qoki61vI5j:b-oqe1an) ([raw file](https://github.com/cockroachdb/cockroach/blob/7850265820cb6e104f2bd6af14e2bb169cb0ded1/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L18)):*
> ```
> 
> statement ok
> CREATE INDEX on t2 (x)
> ```

[nit]: I think you can also remove creation of this index if `x` is defined as `PRIMARY KEY` in the schema.


<!-- Sent from Reviewable.io -->

georgeutsin reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 272 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-Lc90EhLD7_FiJvn4V2D:-LcCh17c51tllk38HZWz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/03b226e87206ffa545c6807265d4576e6955ceff/pkg/sql/exec/mergejoiner_tmpl.go#L272)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Oh, I think I get it. Please leave a comment to describe the purpose of this if/else.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcCKCXsA0qoki61vI5i:-LcCgsfy0z29cAiSViDs:bq45w5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7850265820cb6e104f2bd6af14e2bb169cb0ded1/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L18)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: I think you can also remove creation of this index if `x` is defined as `PRIMARY KEY` in the schema.
</blockquote></details>

Not all the values in `x`are unique, so I can't  make it a primary key.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/distsqlrun/columnar_operators_test.go, line 137 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcBbUnl0y5QrpCJ1lO9:-LcCjFhaE4EEB6O65xGM:byydz86) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ece65752bfa83c6f88186cac48a586df460477f/pkg/sql/distsqlrun/columnar_operators_test.go#L137)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can this generate nulls?
</blockquote></details>

Don't want this comment to get lost as it pertains to this nulls PR. If `MakeRandIntRowsInRange` does not generate nulls we should find a way to do so.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/36727)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @georgeutsin, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/exec_merge_join, line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/36727#-LcCKCXsA0qoki61vI5i:-LcCkRt-1JdAurS0uLog:bco24h9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7850265820cb6e104f2bd6af14e2bb169cb0ded1/pkg/sql/logictest/testdata/logic_test/exec_merge_join#L18)):*
<details><summary><i>Previously, georgeutsin (George Utsin) wrote…</i></summary><blockquote>

Not all the values in `x`are unique, so I can't  make it a primary key.
</blockquote></details>

You can make the non-unique index part of the create table statement like:
```
root@:26257/defaultdb> CREATE TABLE a (x INT, b INT, INDEX x (x));
CREATE TABLE

Time: 5.751ms

root@:26257/defaultdb> INSERT INTO a VALUES (1,2), (1,3);
INSERT 2

Time: 7.526ms

root@:26257/defaultdb>
```
I think the test as it is written now should fail on the insert as `x` is a `PRIMARY KEY` yet you're inserting non-unique values.


<!-- Sent from Reviewable.io -->

georgeutsin commented:
TFTR!

bors r+
